### PR TITLE
fix(agent): stabilize runtime context and generic drafts

### LINKED ISSUE #219: fix(agent): runtime stability — Glob/Grep halt, artifact processing, auto-compact / Agent 运行时稳定性
Tracked from [#210 (comment)](https://github.com/structureclaw/structureclaw/discussions/210#discussioncomment-16746825). Parent issue: #218.

---

## 1.1 `Glob` / `Grep` tools cause Agent output to stop

**Status:** Partially fixed (#212 fixed `grep_files` abort on unsupported files; `Glob` still untested)

**Repro:** Call `Glob` or `Grep` in a long agent session → agent output halts mid-stream.

**Scope:** Verify `Glob` reproduces the same abort; confirm `Grep` fix covers all edge cases (binary files, symlinks, permission errors).

## 1.2 Artifact results unprocessable by Agent

**Status:** Partially fixed (#213 surfaces error details for failed YJK artifacts)

**Scope:** The original issue is broader — the agent cannot interpret *successful* artifact results (e.g., parse analysis output, drive follow-up actions). Error surfacing is done; general artifact consumption is not.

## 1.3 Long-context Agent cannot auto-compact

**Status:** Open

**Impact:** Prolonged conversations exceed model context window → silent truncation or early termination.

**Expected:** Auto-detect context pressure and trigger compact/summarization before hitting the limit.

---

## Progress

| # | Item | Status | Related PRs |
|---|------|--------|-------------|
| 1.1 | `Glob`/`Grep` tools cause Agent stop | ⚠️ Partial | #212 |
| 1.2 | Artifact results unprocessable | ⚠️ Partial | #213 |
| 1.3 | Auto-compact for long context | 🔴 Open | — |

---

---

追踪自 [#210 评论](https://github.com/structureclaw/structureclaw/discussions/210#discussioncomment-16746825)。父 issue：#218。

---

## 1.1 `Glob`、`Grep` 工具导致 Agent 输出停止

**状态：** 部分修复（#212 修复了 `grep_files` 遇到不支持文件时中断的问题；`Glob` 未验证）

**复现：** 在较长的 Agent 会话中调用 `Glob` 或 `Grep` → Agent 输出中途停止。

**范围：** 验证 `Glob` 是否同样会中断；确认 `Grep` 修复覆盖了所有边界情况（二进制文件、符号链接、权限错误）。

## 1.2 `Artifact` 结果 Agent 无法处理

**状态：** 部分修复（#213 为失败的 YJK artifact 暴露了错误详情）

**范围：** 原始问题范围更广——Agent 无法解读*成功的* artifact 结果（例如解析分析输出、驱动后续操作）。错误暴露已完成；通用 artifact 消费尚未实现。

## 1.3 长上下文 Agent 不能自动 compact，对话提前终止

**状态：** 未完成

**影响：** 长时间对话超出模型上下文窗口 → 静默截断或提前终止。

**期望：** 自动检测上下文压力，在达到限制前触发 compact/摘要。

---

## 进度

| # | 项目 | 状态 | 相关 PR |
|---|------|------|---------|
| 1.1 | `Glob`/`Grep` 工具导致 Agent 停止 | ⚠️ 部分完成 | #212 |
| 1.2 | Artifact 结果无法处理 | ⚠️ 部分完成 | #213 |
| 1.3 | 长上下文自动 compact | 🔴 未完成 | — |

### PR DESCRIPTION
**Pull request by qinsz01**
## Summary

Describe the problem and fix in 2-5 bullets:

- Problem: Long agent sessions could stop after workspace search/tool output pressure, analysis artifacts were hard for the model to reason over, context history was not compacted before provider limits, and generic-only structure drafting could discard useful LLM extraction aliases such as `componentType`, `span`, and `pointLoads`.
- Why it matters: Users saw silent/early stops, repeated failed recovery loops, successful analysis results that were too large or too raw to guide follow-up actions, and simple generic-mode requests such as a simply supported beam falling back to `inferredType: unknown`.
- What changed: Hardened `glob_files`/`grep_files`, added context compaction before LLM calls, compacted analysis tool summaries, made draft parameter extraction a direct LLM call with fallback, normalized common generic draft aliases, and rejected empty 0-node/0-element model summaries.
- What did NOT change (scope boundary): This does not add a general artifact interpretation framework for every artifact type; artifact consumption improvements are focused on `run_analysis` summaries. It also does not implement deterministic material/section parsing for generic drafts.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [x] Refactor required for the fix
- [ ] Docs
- [x] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [x] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [ ] Integrations
- [x] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #219
- Related #246
- [x] This PR fixes a bug or regression

## Root Cause / Regression History (if applicable)

- Root cause: Runtime paths assumed tool/model outputs would stay small, well-formed, and directly consumable. Workspace tools did not consistently skip problematic entries, older history was always sent in full, successful analysis artifacts were returned too raw for follow-up reasoning, and the generic structure-type handler only understood `inferredType` instead of common generic aliases returned by the LLM.
- Missing detection / guardrail: No regression tests covered glob/grep symlink/binary/unsupported-file skipping, context compaction preserving active tool protocol, successful analysis artifact summaries, role-safe draft parameter extraction, generic alias normalization, or empty model summaries.
- Prior context (`git blame`, prior PR, issue, or refactor if known): #219 tracks runtime stability follow-up after partial fixes in #212 and #213.
- Why this regressed now: Longer agent sessions and generic-only capability selection expose paths that were previously only lightly exercised: large tool histories, successful artifact follow-up, nested agent message reconstruction, and non-canonical generic LLM extraction shapes.
- If unknown, what was ruled out: The latest simple generic beam failure was not caused by the beam handler; it happened because the user intentionally selected generic and the generic handler discarded alias fields.

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [x] Unit test
  - [x] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file:
  - `backend/src/agent-langgraph/__tests__/workspace-tools.test.mjs`
  - `backend/src/agent-langgraph/__tests__/context-window.test.mjs`
  - `backend/src/agent-langgraph/__tests__/analysis-tool-summary.test.mjs`
  - `backend/src/agent-langgraph/__tests__/param-extractor.test.mjs`
  - `backend/src/agent-skills/structure-type/generic/__tests__/handler.test.mjs`
- Scenario the test should lock in: Search tools skip unsafe/problematic entries instead of aborting; context compaction preserves the current user turn and recent tool protocol; analysis tools return compact actionable summaries; draft extraction avoids nested ReAct role errors; generic mode preserves `componentType/span/pointLoads/supportCondition` as draft state; empty models are reported as failures.
- Why this is the smallest reliable guardrail: These tests target the helper and handler boundaries where the corrupt/oversized state enters the agent loop, without requiring a live LLM provider.
- Existing test that already covers this (if any): Existing chat stream contract validation covers the streaming contract after the helper-level fixes.
- If no new test is added, why not: N/A.

## User-visible / Behavior Changes

- Long conversations can be compacted before model invocation instead of silently overrunning context.
- `glob_files` and `grep_files` now report skip/truncation metadata and avoid aborting on unsupported, binary-looking, symlink, oversized, or unreadable files.
- `run_analysis` tool output is smaller and more actionable: failed runs include compact diagnostics, and successful runs include counts, key metrics, controlling IDs, and warnings.
- Generic-only drafting now accepts common LLM extraction aliases such as `componentType: beam`, `span`, `pointLoads`, and `supportCondition`.
- `build_model` no longer reports a 0-node/0-element model as `success: true`.

## Diagram (if applicable)

```text
Before:
[user asks generic beam] -> extract_draft_params returns aliases -> generic handler drops aliases -> inferredType unknown -> clarification/memory loop
[long session] -> full history/tool output sent to model -> provider/context pressure -> silent stop risk
[analysis success] -> raw artifact-shaped output -> weak follow-up reasoning

After:
[user asks generic beam] -> aliases normalized into draftState -> build_model can proceed
[long session] -> old history summarized + current turn preserved -> LLM call stays bounded
[analysis success/failure] -> compact tool summary + full artifact in graph state -> follow-up reasoning has useful signal
```

## Security Impact (required)

- New permissions/capabilities? (`No`)
- Secrets/tokens handling changed? (`No`)
- New/changed network calls? (`No`)
- Command/tool execution surface changed? (`Yes`)
- Data access scope changed? (`No`)
- If any `Yes`, explain risk + mitigation: Workspace search tools now do stricter bounded traversal and skip symlink/problematic entries. This reduces risk by avoiding traversal into linked external files, binary garbage, very large files, and runaway scans.

### Environment

- OS: Windows / PowerShell
- Runtime/container: local backend Node/TypeScript test environment
- Model/provider: provider-independent unit tests; issue was observed with an OpenAI-compatible provider that rejects empty message roles
- Integration/channel (if any): LangGraph chat agent tools
- Relevant config (redacted): local `settings.json` / LLM config not printed

### Steps

1. Use a long agent session that calls `glob_files` or `grep_files`, including unsupported/binary/symlink-like workspace entries.
2. Run analysis and ask the agent to reason over the result.
3. Continue enough turns to create context pressure.
4. In generic-only mode, ask: `设计一个简支梁，跨度10m，梁中间荷载1kN`.
5. If model building returns empty arrays, inspect the `build_model` tool summary.

### Expected

- Search tools skip problematic entries and return metadata instead of aborting.
- Analysis tool messages are compact but contain success/failure signal and follow-up metrics/diagnostics.
- Older context is compacted while current turn/tool protocol messages remain intact.
- Generic draft aliases are preserved as draft parameters.
- Empty models are reported as failures.

### Actual

- Before this branch, these paths could stop the agent, return oversized/raw artifacts, hit context pressure, lose generic draft aliases, or mark an empty model as successful.
- After this branch, the tests below cover those paths.

## Evidence

Attach at least one:

- [x] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

Passing verification commands:

```bash
npm test --prefix backend -- --runInBand src/agent-langgraph/__tests__ src/agent-skills/structure-type/generic/__tests__/handler.test.mjs
npm run lint --prefix backend
node tests/runner.mjs validate validate-chat-stream-contract
```

Local minimal generic alias check now produces:

```json
{
  "inferredType": "beam",
  "skillId": "generic",
  "structuralTypeKey": "beam",
  "lengthM": 10,
  "supportType": "simply-supported",
  "loadKN": 1,
  "loadType": "point",
  "loadPosition": "midspan",
  "loadPositionM": 5
}
```

## Human Verification (required)

What you personally verified (not just CI), and how:

- Verified scenarios: agent-langgraph test suite, generic handler alias tests, lint, chat stream contract validation, and a direct dist-level generic alias check for the simple supported beam request.
- Edge cases checked: unsupported files, binary-looking files, symlink entries, grep match caps, context compaction preserving current tool protocol, failed analysis diagnostics, successful analysis summaries, draftPatch/direct JSON parsing, explicit `inferredType: "unknown"` not blocking inference from the user message, and empty model summaries.
- What you did **not** verify: Live browser UI flow and a live LLM-driven end-to-end chat run with a real provider after restarting the backend.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

If a bot review conversation is addressed by this PR, resolve that conversation yourself. Do not leave bot review conversation cleanup for maintainers.

## Compatibility / Migration

- Backward compatible? (`Yes`)
- Config/env changes? (`No`)
- Migration needed? (`No`)
- If yes, exact upgrade steps: N/A.

## Risks and Mitigations

- Risk: Context compaction could omit an older detail the model might have used.
  - Mitigation: It only triggers under character pressure, keeps a bounded summary of older turns, and preserves the current user turn plus recent tool protocol messages verbatim.
- Risk: Generic alias normalization could over-infer `beam` from broad text.
  - Mitigation: It only normalizes a small set of common structural aliases and keeps the raw generic draft under `skillState.genericDraft` for auditability.
- Risk: Empty model summaries now return `success: false`, changing downstream behavior that expected success for empty arrays.
  - Mitigation: Empty models are not analyzable; failing early prevents misleading analysis attempts and user confusion.

### CODE REVIEW

File: backend/src/agent-langgraph/tools.ts [hunk 1]
```diff
@@ -173,6 +271,36 @@ export function buildAnalysisToolSummary(args: {
     success: true,
     skillId: args.skillId,
     analysisMode: data?.analysisMode,
+    ...(data ? (buildSuccessfulAnalysisDetails(data, result) ?? {}) : {}),
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-langgraph/tools.ts]: Fixed in `7e5c392`: `buildAnalysisToolSummary` now summarizes successful analysis payloads whether they are under `result.data` or returned at the top level. Added a regression test for top-level `analysisMode`, `summary`, `envelope`, `caseResults`, and `warnings`.

File: backend/src/agent-langgraph/tools.ts [hunk 2]
```diff
@@ -394,14 +522,14 @@ export function createBuildModelTool(skillRuntime: AgentSkillRuntime) {
...
+      const summary = buildModelToolSummary(model, state?.locale || 'zh');
+      const success = summary.success !== false;
+      logToolCall(log, { tool: 'build_model', durationMs: Date.now() - start, success, extra: summary });
       return toolResult(
         toolCallId,
         'build_model',
-        JSON.stringify({ success: true, nodeCount, elementCount, schemaVersion: model.schema_version }),
-        { model },
+        JSON.stringify(summary),
+        success ? { model } : undefined,
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-langgraph/tools.ts]: Fixed in `7e5c392`: empty-model summaries now produce a state update that clears `model` plus downstream `analysisResult`, `codeCheckResult`, and `report`, preventing stale analysis of a previous model. Added a regression test for the cleared state update.

File: backend/src/agent-langgraph/context-window.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+      ];
+
+  const omitted = params.omittedMessageCount > 0
+    ? params.locale === 'zh'
+      ? [`另有 ${params.omittedMessageCount} 条更早消息仅保留在会话检查点中。`]
+      : [`Another ${params.omittedMessageCount} earlier messages remain only in the session checkpoint.`]
+    : [];
+  const lines = params.compactedMessages.map((message) => compactMessageLine(message, params.locale));
+  const content = [...header, ...omitted, '', ...lines].join('\n');
+  return new SystemMessage(truncateText(content, params.summaryCharLimit));
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-langgraph/context-window.ts]: Fixed in `7e5c392`: compacted history is now represented as an `AIMessage` instead of `SystemMessage`, and the summary explicitly says quoted user/tool content is historical data, not new system/developer instructions. Updated context compaction tests to lock this in.

**Review by guyi2000 — APPROVED** [review]
